apps/extension/src/Approvals/ApproveTx/ConfirmTx.tsx

```diff
@@ -0,0 +1,170 @@
+import { useCallback, useEffect, useState } from "react";
+import { useNavigate } from "react-router-dom";
+
+import {
+  Button,
+  ButtonVariant,
+  Input,
+  InputVariants,
+} from "@namada/components";
+import { shortenAddress } from "@namada/utils";
+import { TxType } from "@namada/shared";
+
+import { ApprovalDetails, Status } from "Approvals/Approvals";
+import {
+  ApprovalContainer,
+  ButtonContainer,
+  InfoHeader,
+  InfoLoader,
+} from "Approvals/Approvals.components";
+import { Ports } from "router";
+import { useRequester } from "hooks/useRequester";
+import {
+  SubmitApprovedBondMsg,
+  SubmitApprovedTransferMsg,
+  SubmitApprovedUnbondMsg,
+  SubmitApprovedWithdrawMsg,
+} from "background/approvals";
+import { Address } from "App/Accounts/AccountListing.components";
+import { closeCurrentTab } from "utils";
+import { FetchAndStoreMaspParamsMsg, HasMaspParamsMsg } from "provider";
+import { TxTypeLabel } from "Approvals/types";
+
+type Props = {
+  details?: ApprovalDetails;
+};
+
+export const ConfirmTx: React.FC<Props> = ({ details }) => {
+  const { source, msgId, txType } = details || {};
+  const navigate = useNavigate();
+  const requester = useRequester();
+  const [password, setPassword] = useState("");
+  const [error, setError] = useState<string>();
+  const [status, setStatus] = useState<Status>();
+  const [statusInfo, setStatusInfo] = useState("");
+
+  const handleApproveTx = useCallback(async (): Promise<void> => {
+    setStatus(Status.Pending);
+    setStatusInfo(
+      `Decrypting keys and submitting ${TxTypeLabel[txType as TxType]}...`
+    );
+
+    try {
+      if (!msgId) {
+        throw new Error("msgId was not provided!");
+      }
+      switch (txType) {
```

Comment:
This switch will probably grow right?
I was thinking that maybe instead of the switch we can have a map of TxTypes into Msgs, as those messages have the same constructor parameters. So something like this:
```ts
type SupportedTx = Extract<
  TxType,
  TxType.Bond | TxType.Unbond | TxType.Transfer | TxType.Withdraw
>;

type ApproveMsg = new (msgId: string, password: string) => unknown &
  Message<void>;

const txMap: Map<SupportedTx, ApproveMsg> = new Map([
  [TxType.Bond, SubmitApprovedBondMsg],
  [TxType.Unbond, SubmitApprovedUnbondMsg],
  [TxType.Transfer, SubmitApprovedTransferMsg],
  [TxType.Withdraw, SubmitApprovedWithdrawMsg],
]);
```
We could use them like this:
```ts
const Msg: ApproveMsg | undefined = txMap.get(txType as SupportedTx);

if (Msg) {
    await requester.sendMessage(Ports.Background, new Msg(msgId, password));
}
```
And `SupportedTx` can be removed once we support all of those. 
Not sure if it's more readable, just an idea :)

Reply:
I like this approach better, and this may also be used to clean up one of my utilities.

Reply:
This has been added, replacing the `switch`